core_v14/src/decoder/decode_type.rs

```diff
@@ -0,0 +1,484 @@
+		TypeDefPrimitive::I16 => Primitive::I16(i16::decode(data)?),
+		TypeDefPrimitive::I32 => Primitive::I32(i32::decode(data)?),
+		TypeDefPrimitive::I64 => Primitive::I64(i64::decode(data)?),
+		TypeDefPrimitive::I128 => Primitive::I128(i128::decode(data)?),
+		TypeDefPrimitive::I256 => Primitive::I256(<[u8; 32]>::decode(data)?),
+	};
+	Ok(val)
+}
+
+fn decode_compact_type(
+	data: &mut &[u8],
+	ty: &TypeDefCompact<PortableForm>,
+	types: &PortableRegistry,
+) -> Result<Value, DecodeTypeError> {
+	fn decode_compact(data: &mut &[u8], inner: &Type, types: &PortableRegistry) -> Result<Value, DecodeTypeError> {
+		use TypeDefPrimitive::*;
+		let val = match inner.type_def() {
+			// It's obvious how to decode basic primitive unsigned types, since we have impls for them.
+			TypeDef::Primitive(U8) => Value::Primitive(Primitive::U8(Compact::<u8>::decode(data)?.0)),
+			TypeDef::Primitive(U16) => Value::Primitive(Primitive::U16(Compact::<u16>::decode(data)?.0)),
+			TypeDef::Primitive(U32) => Value::Primitive(Primitive::U32(Compact::<u32>::decode(data)?.0)),
+			TypeDef::Primitive(U64) => Value::Primitive(Primitive::U64(Compact::<u64>::decode(data)?.0)),
+			TypeDef::Primitive(U128) => Value::Primitive(Primitive::U128(Compact::<u128>::decode(data)?.0)),
+			// A struct with exactly 1 field containing one of the above types can be sensibly compact encoded/decoded.
+			TypeDef::Composite(composite) => {
+				if composite.fields().len() != 1 {
+					return Err(DecodeTypeError::CannotDecodeCompactIntoType(inner.clone()));
+				}
+
+				// What type is the 1 field that we are able to decode?
+				let field = &composite.fields()[0];
+				let field_type_id = field.ty().id();
+				let inner_ty = types.resolve(field_type_id).ok_or(DecodeTypeError::TypeIdNotFound(field_type_id))?;
+
+				// Decode this inner type via compact decoding. This can recurse, in case
+				// the inner type is also a 1-field composite type.
+				let inner_value = decode_compact(data, inner_ty, types)?;
+
+				// Wrap the inner type in a representation of this outer composite type.
+				let composite = match field.name() {
+					Some(name) => Composite::Named(vec![(name.clone(), inner_value)]),
+					None => Composite::Unnamed(vec![inner_value]),
+				};
+
+				Value::Composite(composite)
+			}
+			// For now, we give up if we have been asked for any other type:
+			_cannot_decode_from => return Err(DecodeTypeError::CannotDecodeCompactIntoType(inner.clone())),
+		};
+
+		Ok(val)
+	}
+
+	// Pluck the inner type out and run it through our compact decoding logic.
+	let inner =
+		types.resolve(ty.type_param().id()).ok_or_else(|| DecodeTypeError::TypeIdNotFound(ty.type_param().id()))?;
+	decode_compact(data, inner, types)
+}
+
+fn decode_bit_sequence_type(
+	data: &mut &[u8],
+	_ty: &TypeDefBitSequence<PortableForm>,
+	_types: &PortableRegistry,
+) -> Result<BitSequence, DecodeTypeError> {
+	// [jsdw] TODO: might be worth checking the bit_store and bit_order types
```

Comment:
i think the Lsb0/u8 is definition in polkadot-related runtimes, but i parity-scale codec impl Decode/Encode directly on `O/T` generics. Not sure any other project is using anything other than u8/lsb0, and it might significantly complicate bitsequence decoding if we were to make it generic?

Reply:
Yeah; I can see that fairly arbitrary ordering and storage format choices could be made which might be somewhat hard to handle! Offhand it looks like scale-codec just encodes it as the number of bits and then a slice of whatever the underlying storage type is. I _think_ that means that whatever we assume we'll at least decode the correct number of bits, even if we mess up the ordering of them.